semantics-3.0/Semantics.hs

```diff
@@ -52,181 +56,168 @@ data Observation = AndObs Observation Observation
-    ValueLE lhs rhs      -> goVal lhs <= goVal rhs
-    ValueEQ lhs rhs      -> goVal lhs == goVal rhs
-    TrueObs              -> True
-    FalseObs             -> False
-  where
-    goObs = evalObservation env state
-    goVal = evalValue env state
+evalObservation env state obs = let
+    evalObs = evalObservation env state
+    evalVal = evalValue env state
+    in case obs of
+        AndObs lhs rhs       -> evalObs lhs && evalObs rhs
+        OrObs lhs rhs        -> evalObs lhs || evalObs rhs
+        NotObs subObs        -> not (evalObs subObs)
+        ChoseSomething choiceId -> choiceId `Map.member` choices state
+        ValueGE lhs rhs      -> evalVal lhs >= evalVal rhs
+        ValueGT lhs rhs      -> evalVal lhs > evalVal rhs
+        ValueLT lhs rhs      -> evalVal lhs < evalVal rhs
+        ValueLE lhs rhs      -> evalVal lhs <= evalVal rhs
+        ValueEQ lhs rhs      -> evalVal lhs == evalVal rhs
+        TrueObs              -> True
+        FalseObs             -> False
 
 
 -- | Pick the first account with money in it
 refundOne :: Map AccountId Money -> Maybe ((Party, Money), Map AccountId Money)
 refundOne accounts = do
     ((accId, money), rest) <- Map.minViewWithKey accounts
-    if money > 0 then return ((accountOwner accId, money), rest) else refundOne rest
+    if getLovelace money > 0
+    then return ((accountOwner accId, money), rest)
+    else refundOne rest
 
 
 -- | Obtains the amount of money available an account
 moneyInAccount :: AccountId -> Map AccountId Money -> Money
-moneyInAccount = Map.findWithDefault 0
+moneyInAccount = Map.findWithDefault (Lovelace 0)
 
 
 -- | Sets the amount of money available in an account
 updateMoneyInAccount :: AccountId -> Money -> Map AccountId Money -> Map AccountId Money
 updateMoneyInAccount accId money =
-    if money <= 0 then Map.delete accId else Map.insert accId money
+    if getLovelace money <= 0 then Map.delete accId else Map.insert accId money
 
 
 {-| Withdraw up to the given amount of money from an account
     Return the amount of money withdrawn
 -}
 withdrawMoneyFromAccount
   :: AccountId -> Money -> Map AccountId Money -> (Money, Map AccountId Money)
-withdrawMoneyFromAccount accId money accounts = (withdrawnMoney, newAcc)
-  where
-    avMoney        = moneyInAccount accId accounts
-    withdrawnMoney = min avMoney money
-    newAvMoney     = avMoney - withdrawnMoney
-    newAcc         = updateMoneyInAccount accId newAvMoney accounts
+withdrawMoneyFromAccount accId money accounts = let
+    balance        = moneyInAccount accId accounts
+    withdrawnMoney = min balance money
+    newBalance     = Lovelace (getLovelace balance - getLovelace withdrawnMoney)
+    newAcc         = updateMoneyInAccount accId newBalance accounts
+    in (withdrawnMoney, newAcc)
 
 
-{-| Add the given amount of money to an accoun (only if it is positive).
+{-| Add the given amount of money to an account (only if it is positive).
```

Comment:
why can't you add money to an empty account?

Reply:
You can, it is just that it will have no effect if you add an amount <= 0. Empty accounts behave as accounts with 0, that is an optimisation.

Maybe we should add a name for `getLovelace money` since it appears twice.

Reply:
Maybe we should add a name for `getLovelace money` since it appears twice.

Reply:
It would indeed simplify the semantics in some places. But then, by representing both the same way, we are making sure that they behave the same way. It also allows us to have and prove: "accounts always have positive amounts of money" as an invariant.

Reply:
Why aren't we using natural numbers for this, I can't remember. Surely it would simplify a few things nicely.

Reply:
I argue for explicit simplicity versus cleverness. 
We could do plenty of clever tricks there, but what's the purpose?
It is semantics of the language written in Haskell. I believe, that codified semantics is better, and more precise than textual specification. But it must be readable as close to text as sensibly possible.
That's why I'm trying to use as simple Haskell as reasonable to reduce cognitive load on a reader.

Reply:
I get that, it just turned out that a few of these things made the purescript version pretty messy and therefore difficult to understand, this is one of those (and it appears in quite a few places). All the wrapping and unwrapping really obscures what is going on and I think is a product of the tools we want to use (the types) rather than the semantics itself.

Reply:
2 more points:
* Aren't natural numbers simpler?
* An account _is_ a natural number, even if we choose to represent it as an `Integer`, that's just an approximation and it has lead us to write lots of code to deal with the fact that we are not using the actual type (natural numbers). Isn't it clearer to the reader to represent money as natural numbers?

Reply:
Not really. What's the result of ```(Ada 10) - (Ada 20) + (Ada 50)``` in case we use Nat?

Reply:
`Nat`s are not simpler _in Haskell_. 
There is no _standard_ `Nat`, only from 3rd party libraries.
We'll lose ability to implement `Num` for `Nat`s, because naturals do not form a ring (how would you define `negate`?).

All we need from account Id's is that those form a linear order. 
Integers are good enough.
`Text` is worse (because of potential complexities with lexicographical order) but at least it's descriptive.
`Nat` gives no advantages over `Integer`, so we won't gain anything.

So I'm not convinced, basically :)

Reply:
Ah I see, in PS `Nat` implements `Semiring` and has a specific function for dealing with negation (which happens to do what we do). We would have to create our own `Nat` type, not worth it in this case.

It's a bit naughty that all the Haskell `Nat` implementations choose to implement `Num` when they really shouldn't.

Reply:
Well, I can understand this decision considering you don't have `Semiring` in Haskell. 
But you know, not cool :)